### DISCUSSION TITLE
Memory leak : proper way to dispose Avalonia.Media.Imaging.Bitmap after render using IObservable<Bitmap?> binding

### DISCUSSION THREAD

**Discussion by aboccag** [AvaloniaUI/Avalonia#7669] [Q&A] [answered]
Hello everyone ! 
Our team very enjoy the product so far ! Great job.

We are currently developping a computer vision software using Avalonia as a frontend UI.
I have an industrial camera with an output of 60 images per second.

I have a simple app where I just want to display the output stream using the Image control in Avalonia.controls that is using Avalonia.Imaging.Bitmap underthehood. 

The control is as follow : 
```xaml
    <StackPanel>
        <Image Source="{Binding ImageObservable^}"
               Width="720"
               Height="540"
               Name="MyImage" 
        />
    </StackPanel>
```

The ImageObservable is transforming a stream of SKBitmap into a stream of Avalonia.Media.Imaging.Bitmap like this :

```c#

HotImageObservable = new Subject<ICameraResult<T>>();
_camera?.ImagesObservable.Multicast(HotImageObservable).Connect();

        private IObservable<Avalonia.Media.Imaging.Bitmap?> ImageObservable => HotImageObservable
                .Select(img =>
                {
                    using var bitmap = img.Image as SKBitmap;
                    if (bitmap is null) return default;
                    using var stream = bitmap.AsStream();
                    var avaloniaBitmap = new Avalonia.Media.Imaging.Bitmap(stream);
                    return avaloniaBitmap;
                });

```

This code is working but we noticed the memory drastically increase because the var bitmap = new Bitmap(stream); is not disposed.

We are sure that this comes from the new Bitmap(stream) that is not disposed, in fact, the following code shows how I tested to dispose the bitmap inside the stream and there is no memory leak. Obviously no images are displayed.

```c#
private IObservable<Avalonia.Media.Imaging.Bitmap?> ImageObservable => HotImageObservable
                .Select(img =>
                {
                    using var bitmap = img.Image as SKBitmap;
                    if (bitmap is null) return default;
                    using var stream = bitmap.AsStream();
                    using var avaloniaBitmap = new Avalonia.Media.Imaging.Bitmap(stream);
                    return default(Bitmap);
                });

```

The question is: How properly dispose the bitmap after render

Thank you for your help

**Comment by kekekeks** [comment, marked as answer]
I'd suggest to not recreate images but to use WritableBitmap instead.

**Reply by kekekeks** [reply]
Answering the question: it's unclear who is the owner of the Bitmap, but it's probably the observable itself. In that case I'd suggest to retain 2 previous bitmaps in a field and dispose the oldest one. That way you'll only have 2 bitmaps in the memory: one currently held by the `Image` and another already constructed by observable but not yet passed to the `Image`.

**Reply by aboccag** [reply]
> I'd suggest to not recreate images but to use WritableBitmap instead.

I did the following, it works the same but do not resolve the memory leak. Going to try do store the previous reference and dispose it

```c#

        private IObservable<Bitmap?> ImageObservable => HotImageObservable
            .Select(img =>
            {
                using var bitmap = img.Image as SKBitmap;
                if (bitmap is null) return default;

                using var skImage = SKImage.FromBitmap(bitmap);
                using var peakPx = skImage.PeekPixels();
                using var data = peakPx.Encode(SKEncodedImageFormat.Jpeg, 100);
                using var stream = data.AsStream();

                var writeableBitmap = WriteableBitmap.Decode(stream);
                return writeableBitmap;

            })
```

**Reply by aboccag** [reply]
> Answering the question: it's unclear who is the owner of the Bitmap, but it's probably the observable itself. In that case I'd suggest to retain 2 previous bitmaps in a field and dispose the oldest one. That way you'll only have 2 bitmaps in the memory: one currently held by the `Image` and another already constructed by observable but not yet passed to the `Image`.

Seems to work thank you!
